java-manta-client-unshaded/src/main/java/com/joyent/manta/config/AuthAwareConfigContext.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright (c) 2017, Joyent, Inc. All rights reserved.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.joyent.manta.config;
+
+import com.joyent.http.signature.Signer;
+import com.joyent.http.signature.ThreadLocalSigner;
+import com.joyent.http.signature.apache.httpclient.HttpSignatureAuthScheme;
+import org.apache.commons.lang3.BooleanUtils;
+import org.apache.http.auth.Credentials;
+import org.apache.http.auth.UsernamePasswordCredentials;
+
+import java.security.KeyPair;
+import java.util.Objects;
+
+/**
+ * Object for tracking configuration changes related to authentication and recreating dependent objects as needed.
+ * It combines the users' configuration with the derived runtime objects needed to authenticate requests.
+ * objects like the {@link ThreadLocalSigner} which needs careful lifecycle management.
+ *
+ * @author <a href="https://github.com/tjcelaya">Tomas Celayac</a>
+ * @since 3.1.7
+ */
+public class AuthAwareConfigContext
```

Comment:
This class appears to have inconsistent synchronization.  For example, writes to `keyPair` are `synchronized` but reads or not, so a threads may see the updated fields at different times (or never!) and different orders.

Reply:
Created an inner class that holds the derived objects and sits inside an `AtomicReference`. Please review the new behavior.